blog/content/post/2024-11-26-measured-boot-improvements.md

```diff
@@ -0,0 +1,222 @@
+The way the TPM drivers were written in coreboot, did not allow the TPM 1.2
+and TPM 2.0 API to coexist in a single build. So before we approached
+improving the measured boot, we had to make TPM support in coreboot more
+flexible. We have sent and merged a total of 19 (15 if not counting reverted
+commits and their reverts) patches to upstream coreboot, which refactor the
+TPM drivers to allow probing for the TPM device and hooking up the correct
+driver and API to communicate with it.
+
+The list of the patches can be found
+[here](https://review.coreboot.org/q/topic:tpm-unification). A short
+statistic of total lines of code changed (reverted commits and their
+respected reverts excluded):
+
+- 971 deletions
+- 1155 additions
+
+Having the dynamic support for multiple TPM families opened the door to a
+flexible TPM event log compliant with TCG specification.
+
+## Adding TCG-compliant event logging to coreboot
+
+There is no point in the event log if the OS and software cannot
+use/understand it. Trusted Computing Group (TCG) has defined the format of the
+TPM event log for both TPM1.2 and TPM2.0 families. However, coreboot had its
+[own format of
+logging](https://doc.coreboot.org/security/vboot/measured_boot.html#coreboot-specific-format)
+the measurements since the beginning of measured boot support in coreboot. To
+extract the value from the measured boot in coreboot, it was necessary to make
+the event logging compliant with the specifications used throughout the whole
+computer industry. We have sent and merged a total of 10 patches to upstream
+coreboot, adding the TCG-compliant event logging to coreboot's TPM drivers.
+There were also some inconsistencies in naming the event log, e.g. the term
+TCPA, which refers to the TPM 1.2 ACPI table or TPM1.2 event log, was used
+throughout the whole coreboot project to refer to the coreboot's custom TPM
+event log. Now the event log format can be selected during build time between:
+
+- coreboot-specific format
+- TCG TPM1.2 format
+- TCG TPM2.0 format
+
+The list of the patches can be found
+[here](https://review.coreboot.org/q/topic:tpm-std-logs). A short
+statistic of total lines of code changed:
+
+- 360 deletions
+- 1675 additions
+
+Having the proper TPM event logging in coreboot made it easier to be parsed
+and used by the coreboot payloads.
+
+## TPM event log from coreboot to payload
+
+### Problem statement
+
+coreboot alone does not create a usable solution. It focuses merely on proper
+silicon initialization. The functionality to boot an OS is left to the
+application called payload. Payload is a piece of software coreboot hands over
+the execution to, after it is done with all its jobs to initialize the silicon
+and platform. It may take many forms:
+
+- [SeaBIOS](https://www.seabios.org/SeaBIOS) - for legacy BIOS compatibility
+- Linux payload/LinuxBoot/heads - a Linux kernel with optional initrd launched
+ directly by coreboot
```

Comment:
Link to Heads?

Reply:
ade8a4d823b46aa5642a83b03be7b36a62b2c808